### ISSUE TITLE
createSimCode fails in Buildings models

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#7999]
There are many models in the Buildings library that currently fail with
```
[/var/lib/jenkins/ws/OpenModelicaLibraryTestingWork/OpenModelica/OMCompiler/Compiler/SimCode/SimCodeUtil.mo:762:5-762:146:writable]
Error: Internal error function createSimCode failed 
[Transformation from optimised DAE to simulation code structure failed]
```
for example:

- [Buildings.ThermalZones.Detailed.Validation.BESTEST.Cases6xx.Case620](https://libraries.openmodelica.org/branches/master/Buildings_maint.7.0.x/files/Buildings_maint.7.0.x_Buildings.ThermalZones.Detailed.Validation.BESTEST.Cases6xx.Case620.err)
- [Buildings.HeatTransfer.Windows.BaseClasses.Examples.AbsorbedRadiation](https://libraries.openmodelica.org/branches/master/Buildings_maint.7.0.x/files/Buildings_maint.7.0.x_Buildings.HeatTransfer.Windows.BaseClasses.Examples.AbsorbedRadiation.err)
- [Buildings.Examples.VAVReheat.ASHRAE2006](https://libraries.openmodelica.org/branches/master/Buildings_maint.7.0.x/files/Buildings_maint.7.0.x_Buildings.Examples.VAVReheat.ASHRAE2006.err)
- [Buildings.Air.Systems.SingleZone.VAV.Examples.Guideline36](https://libraries.openmodelica.org/branches/master/Buildings_maint.7.0.x/files/Buildings_maint.7.0.x_Buildings.Air.Systems.SingleZone.VAV.Examples.Guideline36.err)

@mahge would you mind having a look?

**Comment by mahge** [comment]
Okay the [results](https://libraries.openmodelica.org/branches/history/master/2021-11-10%2023:06:29..2021-11-11%2011:51:16.html) are in. 

Unfortunately, Buildings got updated at the same time. So it is not clear from the report which is fixed by which. However, the models listed in this ticket (and others which used to have the same issue) are now simulating mostly albeit some of them seem to trigger quite a few asserts. 

Overall it is not so bad. Out of the 1299 models that pass the Backend, there are now just 5 models that fail at C compilation time and a further 61 that do not pass simulation. 

I will investigate further starting from the C compilation. They might be a FrontEnd issue as well. We will see.

**Comment by sjoelund** [comment]
> Unfortunately, Buildings got updated at the same time. So it is not clear from the report which is fixed by which.

Not much changes:
https://github.com/lbl-srg/modelica-buildings/compare/c5ed6ce5e0a7c39859e3167432a7ecab6940397c..80529aa6db22f7aed9df1a38e700b481166ee2bc
https://github.com/lbl-srg/modelica-buildings/compare/4c4e229325c30a47d2f202f183c04170404f9bf9..f9749494688e94ea05924ddb1aaa8bece74f1b9a

**Comment by perost** [comment]
> Overall it is not so bad. Out of the 1299 models that pass the Backend, there are now just 5 models that fail at C compilation time and a further 61 that do not pass simulation.
> 
> I will investigate further starting from the C compilation. They might be a FrontEnd issue as well. We will see.

`Buildings.Controls.DemandResponse.Examples.ClientLBNL90` and `Buildings.Controls.Predictors.Examples.BESTEST` are due to #8120, which are only ones that currently fail at C compilation in 7.0.x.

**Comment by mahge** [comment]
@perost I guess it is a bit tricky to handle the 0 size arrays in practice. 

I have attempted to add a bit more to what you started. The example in this ticket now passes Backend and Codegen but fails at compilation time because `x` is not found, of course. 

We need to think about it a bit more.

**Comment by perost** [comment]
> @perost I guess it is a bit tricky to handle the 0 size arrays in practice.
> 
> I have attempted to add a bit more to what you started. The example in this ticket now passes Backend and Codegen but fails at compilation time because `x` is not found, of course.
> 
> We need to think about it a bit more.

Yeah, I could easily get it to pass the backend, but then the 0-size array variables were just removed and we got the same issue anyway. The new backend is supposed to handle arrays better, and I don't think a general fix for this issue is urgent right now.

**Comment by mahge** [comment]
> I don't think a general fix for this issue is urgent right now.

Ohh. I thought this was needed to simulate the Buildings models. I just checked `Buildings.Controls.DemandResponse.Examples.ClientLBNL90` and it does compile just fine (it is terminated at simulation time but that is another issue). I am guessing it is as a result of #8117.

Then it is all good. We can revisit it later if needed. Although it does not say much about how much effort is left, there is, technically, just one model actually failing in the test suite after #8119 (`simulation/modelica/jacobian`).

**Comment by perost** [comment]
> > I don't think a general fix for this issue is urgent right now.
> 
> Ohh. I thought this was needed to simulate the Buildings models. I just checked `Buildings.Controls.DemandResponse.Examples.ClientLBNL90` and it does compile just fine (it is terminated at simulation time but that is another issue). I am guessing it is as a result of #8117.

Yes, with #8117 the problematic expressions disappear from the model. It's not a general fix for 0-size arrays, but good enough for now I guess.

**Comment by casella** [comment]
> Okay the [results](https://libraries.openmodelica.org/branches/history/master/2021-11-10%2023:06:29..2021-11-11%2011:51:16.html) are in.
> 
> Unfortunately, Buildings got updated at the same time. So it is not clear from the report which is fixed by which.

My estimate was that at least 50 models were affected by the createSimCode issue. Times 3, since we are testing three version. This gives about 150 out of 172 improvements. To me this is one of the best commits of the year 😃 

> However, the models listed in this ticket (and others which used to have the same issue) are now simulating mostly albeit some of them seem to trigger quite a few asserts.

Yeah, I saw that, currently no idea.
 
> I will investigate further starting from the C compilation. They might be a FrontEnd issue as well. We will see.

@mahge, please open another ticket if you identify a separate issue, so we can keep track of what we've done for @mwetter.

**Comment by mwetter** [comment]
Great progress. I think these changes are due to improvements in
OpenModelica as the latest changes in Buildings were for other purposes.

**Comment by casella** [comment]
Yeah, we had several things brewing for a while, and they all got finished during the last couple of days, you see that if you check the latest reports in the [master history log](https://libraries.openmodelica.org/branches/history/master/). Except the last one, that is 😄

**Comment by mahge** [comment]
> This gives about 150 out of 172 improvements. To me this is one of the best commits of the year 😃

😊. I will take it.  

> please open another ticket if you identify a separate issue, so we can keep track of what we've done for @mwetter.

Will do.
